### PR TITLE
[Apple Silicon] [MLX] Fuse SwitchGLU up_proj + gate_proj 

### PR DESCRIPTION
**Pull request by jlee5814**
## Motivation

Refs #22283. First of three planned PRs.

SGLang's MLX backend runs `up_proj` and `gate_proj` as two separate `gather_qmm` calls in every `SwitchGLU` block. This PR concatenates the two weight tensors along the output dim, replacing both calls with one.

On Qwen3-30B-A3B (48 layers), this eliminates 48 kernel dispatches per forward pass. 

Validates Qwen3-30B-A3B-4bit running through SGLang's MoE pipeline on Apple Silicon.

## Modifications

- `FusedSwitchUpGate`: the concatenated projection module.
- `patch_switch_glu_with_fused_up_gate`: walks the model at load time and replaces each `SwitchGLU` block's separate `up_proj` and `gate_proj` with a single `FusedSwitchUpGate`.
- Disabled by default. Enable with `SGLANG_MLX_FUSE_SWITCHGLU=1`.

## Accuracy Tests

Fused and unfused paths produce identical outputs on Qwen3-30B-A3B-4bit: 

max abs diff = 0.00e+00.

GSM8K eval will be addressed once Mac CI is in place.

## Speed Tests and Profiling

Hardware: M4 Pro 24GB, `iogpu.wired_limit_mb=22000`.

`bench_one_batch` Qwen3-30B-A3B-4bit, bs=1, in=60, out=100, n=5 trials 
each, interleaved order:

```bash
SGLANG_USE_MLX=1 SGLANG_MLX_FUSE_SWITCHGLU=$FUSE python -m sglang.bench_one_batch \
  --model-path mlx-community/Qwen3-30B-A3B-4bit \
  --trust-remote-code --disable-radix-cache \
  --tp-size 1 --batch-size 1 --input-len 60 --output-len 100
```

| Configuration | Decode median | Std |
|---|---|---|
| Unfused | 75.12 tok/s | 1.05 |
| Fused (`SGLANG_MLX_FUSE_SWITCHGLU=1`) | 75.67 tok/s | 0.34 |

Difference: +0.55 tok/s (+0.7%).

Fusion produces no measurable improvement at decode bs=1. This matches @adityavaid's RoPE finding: single-op fusion gains nothing at small batch on MLX because lazy-batched dispatch already amortizes the dispatch cost. Future custom kernels need to attack memory traffic, not dispatch count.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests). Deferred - the test loads real model weights, which sits under `test/registered/`. Mac CI infrastructure being set up by @Kangyan-Zhou + @jonahbernard will close this gap.
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

## AI Usage

The main code logic was written by me. Claude was used to discuss design decisions, review the CUDA reference at `sgl-kernel/csrc/moe/cutlass_moe/w4a8/`, analyze multi-trial benchmarks, and polish the PR description. Everything was verified and understood by me.

---
### CI States

Latest PR Test (Base): :x: [Run #26416830814](https://github.com/sgl-project/sglang/actions/runs/26416830814)
Latest PR Test (Extra): :x: [Run #26416830688](https://github.com/sgl-project/sglang/actions/runs/26416830688)

### CODE REVIEW

**Comment by jlee5814** [comment]
Pushed 3ace7b442:

- Skip non-quantized `SwitchGLU`; `FusedSwitchUpGate` assumes `QuantizedSwitchLinear` fields.
- `fused_call` falls back to the original `__call__` for unpatched instances.
- Replaced `assert` with `raise ValueError`.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        do_sort = indices.size >= 64
+        idx = indices
+        inv_order = None
+        if do_sort:
+            x, idx, inv_order = _gather_sort(x, indices)
+        if self.training:
+            idx = mx.stop_gradient(idx)
+
+        # FUSED: single gather_qmm replaces up_proj + gate_proj.
+        x_up, x_gate = self.up_gate_proj(x, idx, sorted_indices=do_sort)
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py]: Could you please check if there is a better way than to duplicate the code from `mlx-lm`?
I think we only need to replace: https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/models/switch_layers.py#L188-L189 with our fused kernel.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            corresponding to what up_proj and gate_proj would have returned.
+        """
+        # Single dispatch instead of two.
+        x_concat = mx.gather_qmm(
+            x,
+            self["weight"],
+            self["scales"],
+            self.get("biases"),  # None-safe via dict-style access
+            rhs_indices=indices,
+            transpose=True,
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py:158]: Please check whether we can pre-transpose the weights/scales in `__init__` and store them in a contiguous layout, so the fused `gather_qmm` path can avoid the runtime transpose logic (`transpose=True`).

Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py:158]: Will benchmark. If `transpose=True` is free in `gather_qmm`, no change; otherwise, I'll pre-transpose in `__init__`.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py (hunk 1 above)
Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py]: Thank you. I'll change `up_proj` and `gate_proj` directly so `SwitchGLU.__call__` is left alone.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    before gate_proj, if that order ever changes, the cache is empty
+    and we raise rather than return None.
+    """
+
+    def __init__(self, up_proxy: "_FusedUpProxy"):
+        self._up_proxy = up_proxy
+
+    def __call__(self, x, indices, sorted_indices=False):
+        gate = self._up_proxy._cached_gate
+        self._up_proxy._cached_gate = None
```
Comment (yeahdongcn) [inline python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py]: Very clever!

Please make sure this follows the Python naming conventions and best practices.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py (hunk 1 above)
Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py]: Pushed `5711a3d5f`. `SwitchGLU.__call__` left alone. Paired proxies on `up_proj` and `gate_proj`: up-proxy runs the fused `gather_qmm` and caches the gate half; gate-proxy returns the cache.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py (hunk 2 above)
Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py:158]: Read `mlx/ops.cpp` and `quantized.metal`:

| mode                    | w packed          | scales          | groups |
|------------------|----------------|-------------|--------|
| transpose=True   | (E, N, K·b/32)   | (E, N, K/g)    | along K |
| transpose=False  | (E, K, N·b/32)   | (E, K, N/g)    | along N |

Groups tied to the last packed axis. Switching modes requires dequantize → requantize, which changes accuracy.

Dispatch is `affine_gather_qmm_t`, tile (bm=16, bn=32, bk=32, wm=1, wn=2). Decode bs=1 with top_k=8 gives M_effective=8 against a 16-row tile. PR 2's leverage.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py (hunk 3 above)
Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py]: Pushed `ea81a5297`. Now goes through `take_cached_gate()` on the up-proxy.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+class _FusedUpProxy:
+    """Drop-in replacement for `switch_mlp.up_proj`.
+
+    Runs the fused gather_qmm, caches the gate half locally, returns just
+    the up half. SwitchGLU.__call__ sees a normal callable returning a single tensor.
+
+    Caching contract:
+        After __call__, self._cached_gate holds the gate half from the most recent fusion.
+        The paired _FusedGateProxy reads and clears it. Single-use: one up_proj call,
+        one gate_proj call, then cleared. Concurrent invocations on the same instance are unsafe;
```
Comment (changminbark) [inline python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py:182]: It might be good to make a TODO to add concurrent support (so someone can take it up in the future) for concurrency support.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+
+        # Build the fused module and the up-proxy.
+        fused = FusedSwitchUpGate(switch_mlp.up_proj, switch_mlp.gate_proj)
+        up_proxy = _FusedUpProxy(fused)
+
+        # Install paired proxies. SwitchGLU.__call__ will invoke up_proj first
+        # (running the fused gather_qmm and caching the gate half on the
+        # up-proxy), then gate_proj (reading the cache from the up-proxy).
+        switch_mlp.up_proj = up_proxy
+        switch_mlp.gate_proj = _FusedGateProxy(up_proxy)
```
Comment (changminbark) [inline python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py]: nit: instead of using `up_proxy`, can you directly create `switch_mlp.up_proj` and then use that reference for the `_FusedGateProxy`?

**Review by changminbark — COMMENTED** [review]
Have you tried running other configurations of `bench_one_batch` like changing batch size and tests like `bench_offline_throughput`? I'm curious what the performance impact is like on those different scenarios.

**Review by alexnails — COMMENTED** [review]
given PR desc can you test bs16 (if needed, scale up to 64 if no measurable difference) and generate a profile?

**Comment by alexnails** [comment]
I’ll do code review either tomorrow or Monday

**Comment by jlee5814** [comment]
@changminbark @alexnails Running bs=16 and bs=64 trials now (covers `bench_offline_throughput` and the profile ask).

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py (hunk 4 above)
Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py:182]: Adding TODO.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py (hunk 5 above)
Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py]: Will inline.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py (hunk 4 above)
Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py:182]: Pushed `a5c45f290`. Added TODO to the caching contract docstring.

File: python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py (hunk 5 above)
Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/fused_switch_glu.py]: Pushed `a5c45f290`. Inlined.

**Comment by jlee5814** [comment]
@changminbark @alexnails Results:

| Config | bs=1 | bs=8 |
|---|---|---|
| Unfused | 75.12 tok/s | 146.40 tok/s |
| Fused | 75.67 tok/s | 143.32 tok/s |
| Δ | +0.7% | -2.1% |

5 trials per config, interleaved. Both within trial variance.

bs=16, bs=64, and `bench_offline_throughput` didn't fit on 24GB; as such, profile was skipped.

File: python/sglang/srt/hardware_backend/mlx/model_runner.py [hunk 6]
```diff
@@ -195,6 +196,17 @@ def _load_model(self):
         load_time = time.time() - start_time
         logger.info(f"MLX model loaded in {load_time:.2f}s")
 
+        # Optional: fuse SwitchGLU's up_proj + gate_proj into one gather_qmm
```
Comment (adityavaid) [inline python/sglang/srt/hardware_backend/mlx/model_runner.py:199]: I would suggest using this as default with fallback and warning info. Whole point of fusion is we use it to improve inference where we can.

File: python/sglang/srt/hardware_backend/mlx/moe/tests/test_fused_switch_glu.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
+"""Numerical equivalence test for FusedSwitchUpGate.
+
+Loads a small MoE model, runs one MoE block both ways (original + patched),
+and asserts outputs match within fp16 tolerance.
+"""
+
+import os
+
```
Comment (adityavaid) [inline python/sglang/srt/hardware_backend/mlx/moe/tests/test_fused_switch_glu.py:8]: pytestmark is applied after importing mlx.core and fused_switch_glu, so a non-Apple job without mlx installed will error before the skip runs.

You can use `pytest.importorskip`, or move MLX imports inside the test after the env-gated skip.

Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/moe/tests/test_fused_switch_glu.py:8]: Pushed `1e2ceb9bd`. Added `pytest.importorskip("mlx.core")` and moved `fused_switch_glu` import inside the test body.

File: python/sglang/srt/hardware_backend/mlx/model_runner.py (hunk 6 above)
Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/model_runner.py:199]: Agreed. bs=16+ OOM's on 24GB, so default-on lacks evidence. @alexnails is running a sweep; will revisit when those numbers land. Opt-in for now, matches #24907.

Comment (jlee5814) [reply python/sglang/srt/hardware_backend/mlx/model_runner.py:199]: Update on the default-on question.

@alexnails ran 5 trials x {fuse=0, fuse=1} x {bs=16, bs=32} paired sweep on Qwen3-30B-A3B-4bit:

Image

All 10 paired trials favor fuse=0. Total throughput and prefill also regress.

Holding opt-in, as no batch size tested shows a positive result on this MoE.

**Comment by jlee5814** [comment]
Closing in favor of Path B (#26188).

Bench data on the targeted Qwen3-30B-A3B-4bit and Qwen1.5-MoE-A2.7B-4bit shows Path B's fused-activation approach is competitive at BS=1 and wins at BS>=4 whereas Path A's weight-concatenation approach regresses at the same shapes.
